### DISCUSSION TITLE
Provide an API to request napari plugins information

### DISCUSSION THREAD

**Discussion by goanpeca** [chanzuckerberg/napari-hub#133] [Ideas]
It would be great if hub provides and API napari could query to get information on plugins without having to query pypi directly.

This would allow to display information in the plugin/package manager inside napari that matches the look feel and information of napari hub.

**Comment by liu-ziyang** [comment]
we actually had an API that was originally designed to be internal use for napari hub, and we do have it in mind to open it up, a few things we need to tackle before opening it up:
1. potential traffic hikes that might take the API down
2. because of this ^ we can't commit to API availability or response time immediately
3. internal review of data flow, even though all data is public, there is personally identifiable information(pii) in these datasets like names and emails, which requires an internal review process for us to go throw to get legal and security approval.

**Comment by goanpeca** [comment]
>potential traffic hikes that might take the API down

That would be an awesome problem to have... ;-) 

Thanks for the feedback @ziyangczi

**Comment by goanpeca** [comment]
I do have question, is the plan to keep a curated set of information on a privately owned server? or is the plan to always ping the pypi API when users arrive to napari hub?

**Comment by liu-ziyang** [comment]
> I do have question, is the plan to keep a curated set of information on a privately owned server? or is the plan to always ping the pypi API when users arrive to napari hub?

we consider PyPI to be the ground truth, but we do cache the result on our AWS s3 to avoid flooding PyPI with traffic, in the future, we may introduce more levels of caching to improve performance

**Comment by neuromusic** [comment]
Love this suggestion, @goanpeca!

Agreed that this would be great to have for the napari viewer instead of napari needing to inspect the pypi.org HTML everytime. Especially since we'll be aggregating non-PyPI data from Github repos, as well, it will be nice to make sure such metadata is available through the napari interface.

For our June launch, all search, sort, and filter functionality will be client-side, but if we moved this to the backend and exposed it through the API, napari would be able to offer the same search/filter/sort functionality that we offer on the napari hub.

So in the near term we could have endpoints for (a) the list of available plugins and (b) metadata for each plugin. With time, this could expand to support (c) search/filter/sort queries.

Even though this isn't especially challenging from a technical perspective, we'll need to work through some CZI internal processes for legal and security review. I'll get the ball rolling on that ASAP, but we won't be able to offer this until after the June launch.

**Comment by liu-ziyang** [comment]
I am glad to announce that we have opened up the API at 
`https://api.napari-hub.org/plugins`

full documentation would be available later
